codersdk/toolsdk/toolsdk.go

```diff
@@ -13,47 +15,198 @@ import (
 	"github.com/coder/coder/v2/codersdk/agentsdk"
 )
 
-// HandlerFunc is a function that handles a tool call.
-type HandlerFunc[T any] func(ctx context.Context, args map[string]any) (T, error)
+// Deps provides access to tool dependencies.
+type Deps struct {
+	CoderClient   *codersdk.Client
+	AgentClient   *agentsdk.Client
+	AppStatusSlug string
+}
+
+// HandlerFunc is a typed function that handles a tool call.
+type HandlerFunc[Arg, Ret any] func(context.Context, Deps, Arg) (Ret, error)
 
-type Tool[T any] struct {
+// Tool consists of an aisdk.Tool and a corresponding typed handler function.
+type Tool[Arg, Ret any] struct {
 	aisdk.Tool
-	Handler HandlerFunc[T]
+	Handler HandlerFunc[Arg, Ret]
 }
 
-// Generic returns a Tool[any] that can be used to call the tool.
-func (t Tool[T]) Generic() Tool[any] {
-	return Tool[any]{
+// Generic returns a type-erased version of a TypedTool where the arguments and
+// return values are converted to/from json.RawMessage.
+// This allows the tool to be referenced without knowing the concrete arguments
+// or return values. The original TypedHandlerFunc is wrapped to handle type
+// conversion.
+func (t Tool[Arg, Ret]) Generic() GenericTool {
+	return GenericTool{
 		Tool: t.Tool,
-		Handler: func(ctx context.Context, args map[string]any) (any, error) {
-			return t.Handler(ctx, args)
-		},
+		Handler: wrap(func(ctx context.Context, tb Deps, args json.RawMessage) (json.RawMessage, error) {
+			var typedArgs Arg
+			if err := json.Unmarshal(args, &typedArgs); err != nil {
+				return nil, xerrors.Errorf("failed to unmarshal args: %w", err)
+			}
+			ret, err := t.Handler(ctx, tb, typedArgs)
+			var buf bytes.Buffer
+			if err := json.NewEncoder(&buf).Encode(ret); err != nil {
+				return json.RawMessage{}, err
+			}
+			return buf.Bytes(), err
+		}, WithCleanContext, WithRecover),
```

Comment:
review: this was the best way I could find to maintain type safety while allowing external callers to treat all the tools equivalently.

Reply:
The other option would be to attempt to fill out the `Arg` via reflection, taking a `map[string]any` as "raw" input. Would definitely be more code, but might be a performance optimization in future for the MCP server which already parses the JSON of the tool invocation.